### PR TITLE
updates for driverMetadata

### PR DESCRIPTION
**Pull request by kinarashah**
add roles, refresh action and calculate url based on latest commit

rancher/rancher#21296
https://github.com/rancher/rancher/issues/21310
https://github.com/rancher/types/pull/905

### CODE REVIEW

File: app/role_data.go [hunk 1]
```diff
@@ -65,9 +65,11 @@ func addRoles(management *config.ManagementContext) (string, error) {
 		addRule().apiGroups("management.cattle.io").resources("multiclusterapps", "globaldnses", "globaldnsproviders", "clustertemplates", "clustertemplaterevisions").verbs("create").
 		addRule().apiGroups("project.cattle.io").resources("sourcecodecredentials").verbs("*").
 		addRule().apiGroups("project.cattle.io").resources("sourcecoderepositories").verbs("*").
-		addRule().apiGroups("management.cattle.io").resources("etcdbackups").verbs("get", "list", "watch")
-
-	rb.addRole("User Base", "user-base").addRule().apiGroups("management.cattle.io").resources("preferences").verbs("*").
+		addRule().apiGroups("management.cattle.io").resources("etcdbackups").verbs("get", "list", "watch").
+		addRule().apiGroups("management.cattle.io").resources("rkek8ssystemimages").verbs("get", "list", "watch").
+		addRule().apiGroups("management.cattle.io").resources("rkek8sserviceoptions").verbs("get", "list", "watch").
+		addRule().apiGroups("management.cattle.io").resources("rkeaddons").verbs("get", "list", "watch").
```
[the remark this answers is not in the document]
Comment (kinarashah) [reply app/role_data.go]: Yeah, that's the intended behavior 
will remove `.`

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 2]
```diff
@@ -130,18 +131,49 @@ func (m *MetadataController) refresh(url string, init bool) {
 		return
 	}
 
+	logrus.Infof("driverMetadata: refresh data")
```
Comment (prachidamle) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: lMaybe change the log statement to Debug?

**Comment by prachidamle** [comment]
As long as we do not update the existing CRDs on metadata refresh, we should be hopefully not have errors introduced due to potential parallel refresh calls. Right @kinarashah ?

**Comment by kinarashah** [comment]
CRDs would get updated 
- if the same key has different values (in between RCs only, for final server versions - we just add new keys) 
- if the existLabel change, which would be when rancher upgrades and RKE now has the latest information, we update label on CRD 

I think in either case, multiple refresh would just result in k8s conflict errors, but data wouldn't get incorrect values

**Comment by prachidamle** [comment]
LGTM

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 3]
```diff
@@ -67,8 +70,11 @@ func Register(ctx context.Context, management *config.ManagementContext) {
 	}
 
 	// load values on startup and start ticker
-	m.refresh(settings.RkeMetadataURL.Get(), true)
-	m.initTicker(ctx)
+	url, interval, err := GetSettingValues()
```
Comment (alena1108) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: shall we check for error before executing line 74?

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 4]
```diff
@@ -130,18 +133,49 @@ func (m *MetadataController) refresh(url string, init bool) {
 		return
 	}
 
+	logrus.Debug("driverMetadata: refresh data")
 	if err := m.createOrUpdateMetadata(data); err != nil {
 		logrus.Errorf("driverMetadata %v", err)
```
Comment (alena1108) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: I feel like we have to enhance this error message

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 5]
```diff
@@ -98,27 +103,25 @@ func (m *MetadataController) sync(key string, test *v3.Setting) (runtime.Object,
...
+		url, _, err := GetSettingValues()
+		if err != nil {
+			logrus.Errorf("driverMetadata: error getting settings %v", err)
+		}
+		m.Refresh(url, false)
 	}
 }
 
-func (m *MetadataController) refresh(url string, init bool) {
+func (m *MetadataController) Refresh(url string, init bool) {
```
Comment (alena1108) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: I think we should add an error as a return value. And then let the caller decide on what to do with it.

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 6]
```diff
@@ -130,18 +133,49 @@ func (m *MetadataController) refresh(url string, init bool) {
...
 		return "", 0, err
 	}
-	return settings.RkeMetadataURL.Get(), checkInterval, nil
+	urlData := map[string]interface{}{}
+	if err := json.Unmarshal([]byte(settings.RkeMetadataURL.Get()), &urlData); err != nil {
+		return "", 0, fmt.Errorf("unmarshal err %v", err)
+	}
+	url, ok := urlData["url"]
+	if !ok {
+		return "", 0, fmt.Errorf("url not present %s", settings.RkeMetadataURL.Get())
```
Comment (alena1108) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: do you think we can enhance the error message?

**Comment by alena1108** [comment]
LGTM, can be merged once vendor.conf conflict is resolved

**Pull request merged**